src/librustc_typeck/check/wfcheck.rs

```diff
@@ -46,6 +46,10 @@ impl<'a, 'gcx, 'tcx> CheckWfFcxBuilder<'a, 'gcx, 'tcx> {
         let param_env = self.param_env;
         self.inherited.enter(|inh| {
             let fcx = FnCtxt::new(&inh, param_env, id);
+            if !inh.tcx.features().trivial_bounds {
+                // Called first so that there is nothing in the cache.
+                check_false_global_bounds(&fcx, span, id);
```

Comment:
Why does it matter if there is something in the cache, actually? It shouldn't, right?

Reply:
I'll add some more explanation to this comment. But the reason is that predicates are cached, rather than obligations. The param_env gets emptied for this check but not for the remaining checks, so the usual wf checks will cache any global bounds as true.